### ISSUE TITLE
superpoint support in LEO-NG

### ISSUE THREAD

**Issue by yarden-carmi** [almog-sharoni/QBench-Release#19]
# simple_nms
- where
- elementwise or/and
- masking
```
def simple_nms(scores, nms_radius: int):
    """ Fast Non-maximum suppression to remove nearby points """
    assert(nms_radius >= 0)

    def max_pool(x):
        return torch.nn.functional.max_pool2d(
            x, kernel_size=nms_radius*2+1, stride=1, padding=nms_radius)

```

# remove_borders
- where
- elementwise or/and
- masking
```
def _sp_remove_borders_list(keypoints, scores, border: int, height: int, width: int):
    return list(zip(*[
        remove_borders(k, s, border, height, width)
        for k, s in zip(keypoints, scores)]))
```
```
def simple_nms(scores, nms_radius: int):
    """ Fast Non-maximum suppression to remove nearby points """
    assert(nms_radius >= 0)

    def max_pool(x):
        return torch.nn.functional.max_pool2d(
            x, kernel_size=nms_radius*2+1, stride=1, padding=nms_radius)

    zeros = torch.zeros_like(scores)
    max_mask = scores == max_pool(scores)
    for _ in range(2):
        supp_mask = max_pool(max_mask.float()) > 0
        supp_scores = torch.where(supp_mask, zeros, scores)
        new_max_mask = supp_scores == max_pool(supp_scores)
        max_mask = max_mask | (new_max_mask & (~supp_mask))
    return torch.where(max_mask, scores, zeros)
```

# top_k_keypoints
```
def _sp_top_k_list(keypoints, scores, k: int):
    return list(zip(*[
        top_k_keypoints(kp, sc, k)
        for kp, sc in zip(keypoints, scores)]))
```
```
def top_k_keypoints(keypoints, scores, k: int):
    if k >= len(keypoints):
        return keypoints, scores
    scores, indices = torch.topk(scores, k, dim=0)
    return keypoints[indices], scores
```

# sample_descriptors
```
def _sp_sample_desc_list(keypoints, descriptors, s: int = 8):
    return [sample_descriptors(k[None], d[None], s)[0]
            for k, d in zip(keypoints, descriptors)]
```
```
def sample_descriptors(keypoints, descriptors, s: int = 8):
    """ Interpolate descriptors at keypoint locations """
    b, c, h, w = descriptors.shape
    keypoints = keypoints - s / 2 + 0.5
    keypoints /= torch.tensor([(w*s - s/2 - 0.5), (h*s - s/2 - 0.5)],
                              ).to(keypoints)[None]
    keypoints = keypoints*2 - 1  # normalize to (-1, 1)
    args = {'align_corners': True} if torch.__version__ >= '1.3' else {}
    descriptors = torch.nn.functional.grid_sample(
        descriptors, keypoints.view(b, 1, -1, 2), mode='bilinear', **args)
    descriptors = torch.nn.functional.normalize(
        descriptors.reshape(b, c, -1), p=2, dim=1)
    return descriptors
```

# _sp_extract_keypoints
```
def _sp_extract_keypoints(scores, threshold: float):
    return [torch.nonzero(s > threshold) for s in scores]
```

# _sp_score_at_keypoints
```
def _sp_score_at_keypoints(scores, keypoints):
    return [s[tuple(k.t())] for s, k in zip(scores, keypoints)]
```

# _sp_flip_kpts
```
def _sp_flip_kpts(keypoints):
    return [torch.flip(k, [1]).float() for k in keypoints]
```

# permute
```
        b, _, h, w = scores.shape
        scores = scores.permute(0, 2, 3, 1).reshape(b, h, w, 8, 8)
        scores = scores.permute(0, 1, 3, 2, 4).reshape(b, h*8, w*8)
```

# L2 norm
`descriptors = torch.nn.functional.normalize(descriptors, p=2, dim=1)`

# softmax + bin
`scores = torch.nn.functional.softmax(scores, 1)[:, :-1]
`
